hooks/get_badge.py

```diff
@@ -0,0 +1,100 @@
+                                    - modified_by
+                                    - name
+                                    - modified_at
+                                    - published_at
+                                    - thumbnail
+                                    - publish_description
+                                    - published_by
+                                    - version
+                                    - entity
+
+        :param publish_path:    String
+                                The path to the publish on disk.
+
+
+
+        :returns:               A QPixmap or QColor to use for the badge, if a badge should be
+                                applied, otherwise None.  If a QColor is returned, a circular "dot"
+                                badge will be generated using that color.
+        """
+        # the default implementation always returns None.
+        return None
+
+    def get_work_file_badge(self, work_file_details, work_file_path, **kwargs):
+        """
+        Generate a badge for a work file.
+
+        :param work_file_details:       Dictionary
+                                A dictionary for the work file to generate a badge for containing
+                                the following keys:
+                                    - task
+                                    - modified_by
+                                    - name
+                                    - modified_at
+                                    - entity
+                                    - version
+                                    - thumbnail
+                                    - description
+
+        :param work_file_path:  String
+                                The path to the work file on disk.
+
+
+        :returns:               A QPixmap or QColor to use for the badge, if a badge should be
+                                applied, otherwise None.  If a QColor is returned, a circular "dot"
+                                badge will be generated using that color.
+        """
+        # the default implementation always returns None.
+        return None
+
+    def generate_badge_pixmap(self, badge_color):
+        """
+        Generate a badge QPixmap from a QColor. This hook method is used to generate a badge image
+        when a badge hook returns a QColor. Thus, by overloading this method, it's possible to
+        customize what the generated badges will look like when get_work_file_badge or
+        get_publish_badge return a QColor.
+
+        :param badge_color:     QColor
+                                The color of the badge to generate a pixmap for.
+
+        :returns:               A QPixmap of the badge to be used.
+        """
+        # We want to multiply the color onto the (white) badge_default dot to
+        # generate a nice looking badge.
+        badge = QtGui.QPixmap(":/tk-multi-workfiles2/badge_default.png")
+        painter = QtGui.QPainter()
+        painter.begin(badge)
```

Comment:
Could we update this to
```python
painter.begin(badge)
try:
        painter.setCompositionMode(QtGui.QPainter.CompositionMode_SourceIn)
        painter.fillRect(badge.rect(), badge_color)
finally:
        painter.end()
```

I've seen hardcrashes in the past when `painter.end` is not invoked properly.

Reply:
Good catch, thanks.  ✅